### PR TITLE
[ci] Run analyze with minimum resolved packages

### PR DESCRIPTION
**Pull request by stuartmorgan**
Since our CI always runs on clean projects, they always get the latest resolved versions of dependencies, so won't catch the easy-to-make mistake of introducing use of a new dependency API without updating the minimum version of that dependency to the version that introduced that API.

This adds a new `--downgrade` flag to `analyze` to run `flutter pub downgrade` before analyzing, and adds a new CI step to run in that mode. (This won't catch runtime mistakes, but as with the "legacy analyze" runs this catches the most common type of mistake at a fraction of the CI cost of running all tests again.)

Also fixes all the existing violations flagged by the new step.

Fixes https://github.com/flutter/flutter/issues/108992

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by stuartmorgan** [comment]
@bparrishMines Follow-up from discussion today. It did require minor tool changes because we have to run `downgrade` in every package individually.

File: script/tool/lib/src/analyze_command.dart [hunk 1]
```diff
@@ -32,12 +32,16 @@ class AnalyzeCommand extends PackageLoopingCommand {
         valueHelp: 'dart-sdk',
         help: 'An optional path to a Dart SDK; this is used to override the '
             'SDK used to provide analysis.');
+    argParser.addFlag(_downgradeFlag,
+        help: 'Runs "flutter pub downgrade" before analysis, to verify that '
```
Comment (GaryQian) [inline script/tool/lib/src/analyze_command.dart:36]: nit: remove comma

File: script/tool/lib/src/analyze_command.dart [hunk 2]
```diff
@@ -113,6 +117,12 @@ class AnalyzeCommand extends PackageLoopingCommand {
       return PackageResult.skip('No lib/ directory.');
     }
 
+    if (getBoolArg(_downgradeFlag)) {
+      if (!await _runPubCommand(package, 'downgrade')) {
```
Comment (GaryQian) [inline script/tool/lib/src/analyze_command.dart:121]: From https://dart.dev/tools/pub/cmd/pub-downgrade:
`The dart pub downgrade command creates a lockfile. If one already exists, pub ignores that file and generates a new one from scratch, using the lowest versions of all dependencies.`

Are we okay with the analyze command creating artifacts? If so, we should note it.

Comment (GaryQian) [reply script/tool/lib/src/analyze_command.dart:121]: I guess pub get also does a similar thing, so this should be fine.

File: .cirrus.yml [hunk 3]
```diff
@@ -171,6 +171,14 @@ task:
         # Only analyze lib/; non-client code doesn't need to work on
         # all supported legacy version.
         - ./script/tool_runner.sh analyze --lib-only --skip-if-not-supporting-flutter-version="$CHANNEL" --custom-analysis=script/configs/custom_analysis.yaml
+    # Does a sanity check that plugins pass analysis with the lowest possible
+    # versions of all dependencies. This is to catch cases where we add use of
+    # new APIs but forget to update minimum versions of dependencies to when
+    # those APIs are introduced.
```
Comment (GaryQian) [inline .cirrus.yml:177]: Awesome! This check seems useful.

File: script/tool/lib/src/analyze_command.dart (hunk 1 above)
Comment (stuartmorgan) [reply script/tool/lib/src/analyze_command.dart]: Done.

File: script/tool/lib/src/analyze_command.dart (hunk 2 above)
Comment (stuartmorgan) [reply script/tool/lib/src/analyze_command.dart:121]: Yes, analyze always creates a lock file because we always have to fetch packages.

**Review by ditman — APPROVED** [review]
This is a very cool new check! Ship it!

**Review by GaryQian — APPROVED** [review]

**Review by bparrishMines — APPROVED** [review]
LGTM

**Pull request merged**